### PR TITLE
fix(haskell): update queries and tests for rewritten parser

### PR DESCRIPTION
**Pull request by mrcjkb**
#6577, but based on `master`, as requested in https://github.com/nvim-treesitter/nvim-treesitter/pull/6577#issuecomment-2094883010

### CODE REVIEW

**Comment by clason** [comment]
Ok, that looks better. Still some failure in folds?

**Comment by mrcjkb** [comment]
@clason it looks like CI is still failing

> invalid node type at position 266 for language haskell

I'll pick this up again in a few days, and see if I can rewrite the queries without the supertypes.

**Comment by clason** [comment]
And please also run `./scripts/format-queries.lua queries/haskell`

**Comment by clason** [comment]
You forgot to bump the lockfile!

**Comment by mrcjkb** [comment]
Still doesn't seem to like the supertype queries.

> Query error at 9:7. Invalid node type "functionpatterns":
(decl/functionpatterns: (patterns

**Comment by clason** [comment]
Yes, anything that goes through Lua is broken by slashes, which need to be escaped in Lua strings.

It's still not clear to me why CI is failing; this needs in-depth debugging.

**Comment by mrcjkb** [comment]
> It's still not clear to me why CI is failing; this needs in-depth debugging.

I believe it's because the formatter changed `decl/function params:` to `decl/functionparams:` (among other things), because of the slashes.

**Comment by clason** [comment]
ah, I see what you mean. yes, that makes sense. @lucario387

**Comment by clason** [comment]
Can you fix the errors manually?

**Comment by mrcjkb** [comment]
There's another issue: The highlight assertions are now picked up as `(haddock)` (documentation string) nodes.

**Comment by mrcjkb** [comment]
Tests are succeeding :smile:

**Comment by clason** [comment]
Ok, then we just need to fix the query formatter. That shouldn't be _too_ hard. (And the query linter, but that will have to wait, and is not a blocker.)

**Comment by lucario387** [comment]
I'm here. Some issues with the pattern: query parser doesn't accept supertypes like that yet.

`(foo/bar)` results in a tree such as

```scheme
(named_node
  name: (identifier) ; foo
  (ERROR
    (ERROR) ; /
    (identifier)) ; bar
```

So I think I need to become less clueless about how to update the query parser first :D

**Comment by clason** [comment]
I don't see what `has-ancestor?` has to do with this?

**Comment by lucario387** [comment]
Am I understanding something wrong or is this for capturing node that has a certain descendant, or a node with a certain ancestor?

**Comment by clason** [comment]
I don't know, and I don't really care ;) Haskell is the only parser using this, and I don't see any other parser rewriting everything in terms of supertype nodes.

TL;DR: has-ancestor? is going nowhere; supertypes mostly seem a niche way of organizing nodes (which has been supported for a loooong time).

**Comment by mrcjkb** [comment]
> Am I understanding something wrong or is this for capturing node that has a certain descendant, or a node with a certain ancestor?

I just tried rewriting

```scheme
(decl/function) @_func
```

as

```scheme
((function) @_func
 (#has-ancestor? @_func decl))
```

and it doesn't seem to be capturing anything.

**Comment by clason** [comment]
Yeah, that's something different. I think the query parser just needs to be taught that `foo/bar` is a valid `identifier`.

**Comment by clason** [comment]
Could it really be this easy?

**Comment by clason** [comment]
@mrcjkb try rebasing on latest master

File: queries/haskell/highlights.scm [hunk 1]
```diff
@@ -4,66 +4,33 @@
...
-(function
-  patterns: (patterns
-    (_) @variable.parameter))
+(decl/function patterns: (patterns
+  (_) @variable.parameter))
 
-(exp_lambda
-  (_)+ @variable.parameter
-  "->")
+(expression/lambda (_)+ @variable.parameter
```
Comment (lucario387) [inline queries/haskell/highlights.scm:12]: Is it missing a run or is it already done? This looks a bit off.

Comment (lucario387) [reply queries/haskell/highlights.scm:12]: nvm the formatter goes even worse

Comment (lucario387) [reply queries/haskell/highlights.scm:12]: Freaking hell. Why does half the time the result is correct and the other half it's completely wrong ??????????

Comment (clason) [reply queries/haskell/highlights.scm:12]: supertypes! \o/

Comment (mrcjkb) [reply queries/haskell/highlights.scm:12]: oof. Sorry :cry:

Comment (clason) [reply queries/haskell/highlights.scm:12]: Nah, that's our problem; supertypes are official, and we'll have to support them sooner or later. We'll get there.

**Comment by mrcjkb** [comment]
[Note to self] Downstream plugins/packages to be updated:

- [ ] [neotest-haskell](https://github.com/mrcjkb/neotest-haskell)
- [ ] [vim-matchup](https://github.com/andymass/vim-matchup)
- [ ] [iswap.nvim](https://github.com/mizlan/iswap.nvim)
- [ ] [rainbow-delimiters.nvim](https://gitlab.com/HiPhish/rainbow-delimiters.nvim)
- [ ] [haskell-scope-highlighting.nvim](https://github.com/kiyoon/haskell-scope-highlighting.nvim)
- [ ] [nvim-treesitter-context](https://github.com/nvim-treesitter/nvim-treesitter-context)
- [ ] [nvim-treesitter-textobjects](https://github.com/nvim-treesitter/nvim-treesitter-textobjects)
- [ ] [tailwind-sorter.nvim](https://github.com/laytan/tailwind-sorter.nvim)
- [ ] [ttags](https://github.com/npezza93/ttags)

**Comment by tchoutri** [comment]
@mrcjkb any plans for neotest-haskell? This update gives me a curious error message:

```
All parsers are up-to-date!
Error detected while processing BufReadPost Autocommands for "*":
Error executing lua callback: /usr/share/nvim/runtime/filetype.lua:35: Error executing lua: /usr/share/nvim/runtime/filetype.lua:36: BufReadPost Autocommands for "*"..FileType Autocommands for "*": Vim(append):Er
ror executing lua callback: /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:252: Query error at 1:3. Invalid node type "exp_apply":
((exp_apply
  ^

stack traceback:
        [C]: in function '_ts_parse_query'
        /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:252: in function 'fn'
        /usr/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'fn'
        /usr/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'get_query'
        ...r-textobjects/lua/nvim-treesitter/textobjects/shared.lua:99: in function 'available_textobjects'
        ...r-textobjects/lua/nvim-treesitter/textobjects/select.lua:173: in function 'attach'
        ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:509: in function 'attach_module'
        ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:532: in function 'reattach_module'
        ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:133: in function <...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:132>
        [C]: in function 'nvim_cmd'
        /usr/share/nvim/runtime/filetype.lua:36: in function </usr/share/nvim/runtime/filetype.lua:35>
        [C]: in function 'nvim_buf_call'
        /usr/share/nvim/runtime/filetype.lua:35: in function </usr/share/nvim/runtime/filetype.lua:10>
stack traceback:
        [C]: in function 'nvim_cmd'
        /usr/share/nvim/runtime/filetype.lua:36: in function </usr/share/nvim/runtime/filetype.lua:35>
        [C]: in function 'nvim_buf_call'
        /usr/share/nvim/runtime/filetype.lua:35: in function </usr/share/nvim/runtime/filetype.lua:10>
stack traceback:
        [C]: in function 'nvim_buf_call'
        /usr/share/nvim/runtime/filetype.lua:35: in function </usr/share/nvim/runtime/filetype.lua:10>
Press ENTER or type command to continue
```

**Comment by tchoutri** [comment]
@mrcjkb any plans for neotest-haskell? This update gives me a curious error message:

```
All parsers are up-to-date!
Error detected while processing BufReadPost Autocommands for "*":
Error executing lua callback: /usr/share/nvim/runtime/filetype.lua:35: Error executing lua: /usr/share/nvim/runtime/filetype.lua:36: BufReadPost Autocommands for "*"..FileType Autocommands for "*": Vim(append):Error executing lua callback: /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:252: Query error at 1:3. Invalid node type "exp_apply":
((exp_apply
  ^

stack traceback:
        [C]: in function '_ts_parse_query'
        /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:252: in function 'fn'
        /usr/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'fn'
        /usr/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'get_query'
        ...r-textobjects/lua/nvim-treesitter/textobjects/shared.lua:99: in function 'available_textobjects'
        ...r-textobjects/lua/nvim-treesitter/textobjects/select.lua:173: in function 'attach'
        ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:509: in function 'attach_module'
        ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:532: in function 'reattach_module'
        ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:133: in function <...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:132>
        [C]: in function 'nvim_cmd'
        /usr/share/nvim/runtime/filetype.lua:36: in function </usr/share/nvim/runtime/filetype.lua:35>
        [C]: in function 'nvim_buf_call'
        /usr/share/nvim/runtime/filetype.lua:35: in function </usr/share/nvim/runtime/filetype.lua:10>
stack traceback:
        [C]: in function 'nvim_cmd'
        /usr/share/nvim/runtime/filetype.lua:36: in function </usr/share/nvim/runtime/filetype.lua:35>
        [C]: in function 'nvim_buf_call'
        /usr/share/nvim/runtime/filetype.lua:35: in function </usr/share/nvim/runtime/filetype.lua:10>
stack traceback:
        [C]: in function 'nvim_buf_call'
        /usr/share/nvim/runtime/filetype.lua:35: in function </usr/share/nvim/runtime/filetype.lua:10>
Press ENTER or type command to continue
```

**Comment by mrcjkb** [comment]
> @mrcjkb any plans for neotest-haskell? This update gives me a curious error message:
> 
> ```
> All parsers are up-to-date!
> Error detected while processing BufReadPost Autocommands for "*":
> Error executing lua callback: /usr/share/nvim/runtime/filetype.lua:35: Error executing lua: /usr/share/nvim/runtime/filetype.lua:36: BufReadPost Autocommands for "*"..FileType Autocommands for "*": Vim(append):Er
> ror executing lua callback: /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:252: Query error at 1:3. Invalid node type "exp_apply":
> ((exp_apply
>   ^
> 
> stack traceback:
>         [C]: in function '_ts_parse_query'
>         /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:252: in function 'fn'
>         /usr/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'fn'
>         /usr/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'get_query'
>         ...r-textobjects/lua/nvim-treesitter/textobjects/shared.lua:99: in function 'available_textobjects'
>         ...r-textobjects/lua/nvim-treesitter/textobjects/select.lua:173: in function 'attach'
>         ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:509: in function 'attach_module'
>         ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:532: in function 'reattach_module'
>         ...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:133: in function <...vim/lazy/nvim-treesitter/lua/nvim-treesitter/configs.lua:132>
>         [C]: in function 'nvim_cmd'
>         /usr/share/nvim/runtime/filetype.lua:36: in function </usr/share/nvim/runtime/filetype.lua:35>
>         [C]: in function 'nvim_buf_call'
>         /usr/share/nvim/runtime/filetype.lua:35: in function </usr/share/nvim/runtime/filetype.lua:10>
> stack traceback:
>         [C]: in function 'nvim_cmd'
>         /usr/share/nvim/runtime/filetype.lua:36: in function </usr/share/nvim/runtime/filetype.lua:35>
>         [C]: in function 'nvim_buf_call'
>         /usr/share/nvim/runtime/filetype.lua:35: in function </usr/share/nvim/runtime/filetype.lua:10>
> stack traceback:
>         [C]: in function 'nvim_buf_call'
>         /usr/share/nvim/runtime/filetype.lua:35: in function </usr/share/nvim/runtime/filetype.lua:10>
> Press ENTER or type command to continue
> ```

Yep, it's the first on my list: https://github.com/mrcjkb/neotest-haskell/pull/162

Until that's merged, it'll only work with the old parser, which means you would have to pin nvim-treesitter to a commit from before this PR.
